### ISSUE TITLE
Getting ECS to 1.0

### ISSUE THREAD

**Issue by webmat** [elastic/ecs#115]
# Getting ECS to Beta1

Here's an attempt at summing the work that should be completed, in order to get ECS to Beta1.

## Why this plan?

* We need to get agreement on what should be in 1.0, and what is kept for later
* We need to have this plan in writing, so we can set expectations on timeline and amount of effort remaining
* With everyone slowly ramping up to conform to ECS, we need to make it clear what's still planned to change
* We need to have as much of this ready as early as possible, so that people working to implement ECS are working towards the right thing :-)

## How?

Part of the strategy is dividing the fields into ECS Core (which should be more stable) and ECS Extended (which may change a little more in the future), and making sure at least Core is very stable for 1.0. We'll do our best to stabilize ECS Extended as well, on a best effort basis.

I may have included too much in this plan for 1.0, challenging each issue's inclusion for 1.0 (especially if they're ECS Extended) is welcome ;-)

This issue will contain lots of checkmark tasks. All of these should eventually become a concrete GitHub issue. But this is not the case yet: the idea is to get my view of this plan out there, discuss and firm up the high level plan ASAP. Then we can firm up what needs to happen where, create subsequent tasks or PRs, to get to 1.0.

We should discuss/challenge this plan here, and update the body of the overarching plan (this issue) accordingly.

I think it will be helpful to label all issues that that go into this plan with a label such as `ga-1.0`.

Finally, some of these issues are still written in extreme shorthand (e.g. 1 or 2 words), that may only make sense for the folks closest to the ECS discussions. Feel free to ask for clarifications :-)

Without further ado, here's the plan how I see it at the moment.

### Multiple issues

- This issue is for work considered blockers to declare GA
- #161 is for work related to GA but non-blocker tasks
- #160 is for post-GA tasks

# ECS plan for Beta1

- [x] Perform the actual release (tag, announcements, etc)

## ECS Core/Extended/Other (93)

- [x] Document which field is Core & Extended (#93)
  - Mention the 3rd level as well, right now named "Other". There's discussions around having "Application", "Service", "Protocol"
  - [x] Assign each field to Core or Extended in the yml
  - What do we do with "Other"? Leave unassigned in the yml?
- [x] Update the script generating the doc to indicate Core vs Extended fields (#149)

## Top Level Objects

The "Top Level Objects" section is meant to outline which objects will stay at the top level and which ones are slated to be 1) nested; 2) become a "reuseable" object (nestable in different places); or 3) are new objects.

The work that needs to happen in these TLOs (changes on the fields themselves) is listed in the other sections below.

### Core

- agent
- cloud
- container
- destination
- device
- event
- error
- file
- [x] geo
  - reuseable object
- [x] http
  - take out temporarily (#171)
- host
- [x] kubernetes (removed in #132)
- log
- network
- organization
- [x] os
  - reuseable object
- process
- service
- source
- [x] tls (removed in #132)
- url
- [x] user
  - reuseable object
- [x] user_agent 
  * Move to use cases (#172)

## Structure

- [x] Compose vs Reuse
  - [x] close discussion (#71, #72, #94)
  - [x] Document which objects are reuseable, how and where (#94)
  - [x] reuseable:
    - geo (closes #9)
    - os
    - user
    - threat
  - [x] Remove nested `os` in some tables where it was initially duplicated manually a while ago. Now that it's a reusable object, it's expected in these places, but the duplication of the fields under `user_agent.os` and `host.os` is superfluous (#168)

# Datatypes

- [x] New convention for text indexing
  - (Almost) always `keyword` at the top level (#137)
  - Analyzed will be the nested field if needed. Users can add analyzed nested field without breaking compatibility with ECS. (#140)
  - Two exceptions to this rule, to avoid further breaking changes. The following fields will be `text` only indexing: `message`, `error.message`.
  - [x] Rewrite the doc that talks about multi-fields to reflect the new approach. (#140)
  - [x] We will need to close issues / undo all other multi-fields. See #103, #104, #118, #121

## Naming

- [x] hostname (#31, #62, pr #74)
- [x] url.original (pr #109)

## Documentation

- [x] Clarify use of hostname, subdomain, domain in source/destination (#84, hostname PR #175)
  - Note: subdomain-related fields have been taken out of beta 1
- [x] Adjust the README intro for the release (#167, PR #174)

## Network discussions

- [x] network protocol layers (pr #81)
  - [x] protocol.name and .id (pr #79)

## ECS

- [x] Rename event.version to ecs.version (pr #169)

**Comment by webmat** [comment]
I just updated the issue based on today's ECS meeting.

**Comment by webmat** [comment]
@MikePaquette @ruflin

I've added a new H1 section to put the things related to GA, but not blockers to declaring the schema itself GA.

* The "automation" and "Related projects" H2 sub-sections were already at the end of my list, so they are already in that new section for later.
* I've also created another "documentation" section for the things that aren't that pressing to document either.
* I've moved a few odd tasks that were elsewhere but about "automation" to that new section for after GA

Another thing of note, I've removed the following task: "Review current fields and consider making some longs into ints (e.g. port)"

**Comment by ruflin** [comment]
Can we move everything which is post GA and for other projects into a different issue? So this issue is purely about going GA?

**Comment by webmat** [comment]
@ruflin I had already been thinking about splitting this in 3, actually. This issue's purpose would become to track work for internal availability (IA), create a new one for some more work in getting this ready for GA, and finally one that would be post-ga (but related to GA, close after).

WDYT?

**Comment by ruflin** [comment]
SGTM

**Comment by webmat** [comment]
I've removed a bunch of mentions of "GA" and 1.0 (not all), in favour of "Beta1". Beta1 is what we're shipping as of Stack 7.0.

@ruflin I've moved a few things to later:

Moved to #161:

- Reflect object reuse in template.json
- Reflect object reuse in schema.csv
- Fix script issue with `multi_fields`, introduced when adding for Core/Extended

Moved to #160:

- Introduce `related.*` field set (#67)
- Discuss nesting `http` (and eventually all network protocol field sets) under `network`?

And added a new task:

- Adjust the readme for the release (#167)
